cmd/boostd/genindex.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/filecoin-project/boostd-data/model"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/ipfs/go-cid"
+	carv1 "github.com/ipld/go-car"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/multiformats/go-varint"
+	"github.com/urfave/cli/v2"
+)
+
+var genindexCmd = &cli.Command{
+	Name:  "genindex",
+	Usage: "Generate index for a given piececid and store in the piece directory",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:  "add-index-throttle",
+			Usage: "the maximum number of add index operations that can run in parallel",
+			Value: 4,
+		},
+		&cli.StringFlag{
+			Name:     "api-fullnode",
+			Usage:    "the endpoint for the full node API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "api-storage",
+			Usage:    "the endpoint for the storage node API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "api-piece-directory",
+			Usage:    "the endpoint for the piece directory API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "piececid",
+			Usage:    "piececid to index",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "filepath",
+			Usage:    "the path to the car",
+			Required: false,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+
+		// parse piececid
+		piececid, err := cid.Decode(cctx.String("piececid"))
+		if err != nil {
+			return err
+		}
+		fmt.Println("piececid to index: ", piececid)
+
+		// connect to the piece directory service
+		pdClient := piecedirectory.NewStore()
+		defer pdClient.Close(ctx)
+		err = pdClient.Dial(ctx, cctx.String("api-piece-directory"))
+		if err != nil {
+			return fmt.Errorf("error while connecting to piece directory service: %w", err)
+		}
+
+		// connect to the full node
+		fnApiInfo := cctx.String("api-fullnode")
+		fullnodeApi, ncloser, err := lib.GetFullNodeApi(ctx, fnApiInfo, log)
+		if err != nil {
+			return fmt.Errorf("getting full node API: %w", err)
+		}
+		defer ncloser()
+
+		// connect to the storage API and create a sector accessor
+		storageApiInfo := cctx.String("api-storage")
+		sa, storageCloser, err := lib.CreateSectorAccessor(ctx, storageApiInfo, fullnodeApi, log)
+		if err != nil {
+			return err
+		}
+		defer storageCloser()
+
+		pr := &piecedirectory.SectorAccessorAsPieceReader{SectorAccessor: sa}
+
+		pd := piecedirectory.NewPieceDirectory(pdClient, pr, cctx.Int("add-index-throttle"))
+
+		var r *carv2.Reader
+		var headerDataSize uint64
+
+		if filepath := cctx.String("filepath"); filepath != "" {
+			fmt.Println("fetching piececid from car file: ", filepath)
+
+			r, err = carv2.OpenReader(filepath)
+			if err != nil {
+				return fmt.Errorf("open reader: %w", err)
+			}
+			defer r.Close()
+
+			if r.Version == 1 {
+				fi, err := os.Stat(filepath)
+				if err != nil {
+					return fmt.Errorf("os.stat err: %w", err)
+				}
+				headerDataSize = uint64(fi.Size())
+			}
+		} else {
+			fmt.Println("no car file specified, fetching piececid from the storage api")
```

Comment:
Can we simply expose  `buildIndexForPiece` method in piecedirectory and call it since this would be an internal process?

Reply:
Yes, all this code probably shouldn't belong on the CLI layer.